posts/solving-parallel-letter-frequency-elixir-part-2-adding-concurrency.md

```diff
@@ -0,0 +1,127 @@
+
+[Parallel Letter Frequency](https://exercism.io/tracks/elixir/exercises/parallel-letter-frequency) is a medium difficulty exercise on [Exercism's Elixir Track](https://exercism.io/tracks/elixir) that unpacks a surprising number of interesting lessons. One of those lessons is how easy it is to add concurrency to Elixir code with core features alone, which has clear benefits for anyone writing applications whose performance could be improved by parallelizing tasks. 
+
+This exercise requires you to implement a `Frequency.frequency/2` function that determines letter frequency in a list of strings, with the added challenge of distributing the calculation to a number of worker processes, set with the `workers` argument:
+
+```elixir
+iex> Frequency.frequency(["Freude", "schöner", "Götterfunken"], workers)
+%{
+  "c" => 1, 
+  "d" => 1, 
+  "e" => 5, 
+  ...
+  "ö" => 2
+}
+```
+
+Let's explore how we can add concurrency to our solution using Elixir.
+
+## Writing concurrent Elixir code with the `Task` module
+
+A big selling point of Elixir is its scalability, which comes as a result of all Elixir code running concurrently in [Erlang VM](https://en.wikipedia.org/wiki/BEAM_(Erlang_virtual_machine)) processes. You can very easily spawn processes with functions like [`Kernel.spawn_link/1`](https://hexdocs.pm/elixir/Kernel.html#spawn_link/1), but if concurrency is the goal, you're much better off using the awesomely powerful abstractions provided by the [`Task` module](https://hexdocs.pm/elixir/Task.html):
+
+> The most common use case for [Task] is to convert sequential code into concurrent code by computing a value asynchronously.
+
+The `Task` module lets you write unbelievably clean concurrent code in Elixir -- no [callback hell](http://callbackhell.com/) and no need for [Promises](https://javascript.info/promise-basics). 
+
+For the problem at hand, [`Task.async_stream/5`](https://hexdocs.pm/elixir/Task.html#async_stream/5) is a great option:
+
+```elixir
+async_stream(enumerable, module, function, args, options \\ [])
+```
+
+`Task.async_stream/5` returns a stream that runs the given `module`, `function`, and `args` concurrently on each item in `enumerable`. There are two ways to control the level of concurrency with `Task.async_stream/5`:
+
+1. The number of items in `enumerable`, which determines the number of workers spawned (in the absence of the `:max_concurrency` option)
+2. Setting the `:max_concurrency` option to the number of workers, given that the number of items in `enumerable` is greater than or equal to `:max_concurrency`
+
+The first option is simple and easy to understand, and lets you ensure the correct number of workers is used even for short strings. 
+
+## Make the letter frequency calculation concurrent
```

Comment:
In this section, I wonder if it would help to add a little bit of intro about what is going to happen.

We'll take an existing sequential implementation, and parallelize it by doing [x] (which you've described in the "in words" bullet list). If we do this I think it would make sense to put the diagram after the bullet list, rather than before it.

Reply:
I'll go over this one again and see if I can make it clearer.